pkg/cloud/baremetal/actuators/machine/actuator_test.go

```diff
@@ -1630,3 +1631,208 @@ func TestDeleteOfBareMetalHostDeletesMachine(t *testing.T) {
 		}
 	}
 }
+
+func TestRemediation(t *testing.T) {
+	machine, machineNamespacedName := getMachine("machine1")
+	host, hostNamespacedName := getBareMetalHost("host1")
+	node, nodeNamespacedName := getNode("node1")
+	linkMachineAndNode(machine, node)
+	host.Status.PoweredOn = true
+
+	//starting test with machine that needs remediation
+	machine.Annotations = make(map[string]string)
+	machine.Annotations[HostAnnotation] = host.Namespace + "/" + host.Name
+	machine.Annotations[externalRemediationAnnotation] = ""
+
+	scheme := runtime.NewScheme()
+	clusterapis.AddToScheme(scheme)
+	bmoapis.AddToScheme(scheme)
+	corev1.AddToScheme(scheme)
+
+	c := fakeclient.NewFakeClientWithScheme(scheme)
+
+	actuator, err := NewActuator(ActuatorParams{
+		Client: c,
+	})
+	if err != nil {
+		t.Error(err)
+	}
+
+	c.Create(context.TODO(), machine)
+	c.Create(context.TODO(), host)
+	c.Create(context.TODO(), node)
+
+	err = actuator.Update(context.TODO(), nil, machine)
+	if err != nil {
+		t.Errorf("unexpected error %v", err)
+	}
+
+	host = &bmh.BareMetalHost{}
+	c.Get(context.TODO(), hostNamespacedName, host)
+	if !hasPowerOffRequestAnnotation(host) {
+		t.Log("Expected reboot annotation on the host but none found")
+		t.Fail()
+	}
+
+	host.Status.PoweredOn = false
+	c.Update(context.TODO(), host)
+
+	err = actuator.Update(context.TODO(), nil, machine)
+	if err != nil {
+		t.Errorf("unexpected error %v", err)
+	}
+
+	machine = &machinev1.Machine{}
+	c.Get(context.TODO(), machineNamespacedName, machine)
+
+	if _, exists := machine.Annotations[poweredOffForRemediation]; !exists {
+		t.Log("Expected powered-off-for-remediation annotation to exist on machine but none found")
+		t.Fail()
+	}
+
+	err = actuator.Update(context.TODO(), nil, machine)
+	if err != nil {
+		switch err.(type) {
+			case *clustererror.RequeueAfterError:
```

Comment:
Why do we expect/allow this type of error here?

Reply:
Actuator can tell the controller to requeue by returning this type of error. In this `Update()` we expect the node to be deleted by `deleteNode(...)` function, which request a requeue, and this is expected

Reply:
Should the test fail if err is nil, then? Do we always expect the error or is it just that we allow it?

Reply:
it should fail, fixed that